### DISCUSSION TITLE
Unable to prepare statement with "member of" in query over `pscale connect`

### DISCUSSION THREAD

**Discussion by jvliwanag** [planetscale/discussion#593] [Bug Reports] [answered]
When using `pscale connect` to proxy mysql connections onto planetscale - prepare statements with `member of` fail.

Examples:

```
SQL: select ? member of ('[]')

Error: syntax error at position 19 near 'of'
```

```
SQL: update products set update_time=now() where ? member of (json_keys(option_sets))

Error: syntax error at position 53 near 'member'
```

Note that connecting directly to planetscale bypassing the `pscale connect` proxy works just fine. Also, non prepared statements (direct executions) also work fine.

**Comment by orware** [comment]
Hello @jvliwanag,

I haven't been able to replicate the same issue so far, although your report can still be valid under specific conditions.

I ran a simplified test using the following set of queries:

```sql
PREPARE stmt1 FROM 'SELECT ? MEMBER OF (''[23, "abc", 17, "ab", 10]'') is_member';
SET @a = 17;
EXECUTE stmt1 USING @a;
DEALLOCATE PREPARE stmt1;
``` 

And I ran the above from a MySQL connection in DBeaver first which was successful, then I ran a second test via our `pscale shell` command against the same database which was also successful. Lastly, I used `pscale connect` and then setup a new connection via DBeaver for that option, but the queries above were all still successful so I wasn't able to encounter the same `syntax error` message at least with the simple test above.

Example output when using `pscale shell`:

```
> PREPARE stmt1 FROM 'SELECT ? MEMBER OF (''[23, "abc", 17, "ab", 10]'') is_member';
test/|⚠ main ⚠|> SET @a = 17;
test/|⚠ main ⚠|> EXECUTE stmt1 USING @a;
+-----------+
| is_member |
+-----------+
|         1 |
+-----------+
test/|⚠ main ⚠|> DEALLOCATE PREPARE stmt1;
test/|⚠ main ⚠|> exit;
```

Example output from DBeaver for 1st (direct) and 3rd (via `pscale connect`) tests:
<img width="656" alt="image" src="https://github.com/planetscale/discussion/assets/931650/b9f807b7-fbab-4d46-93ca-303b76a6038c">

If this is something you would like to try and troubleshoot a little further with us however please send in a ticket via https://support.planetscale.com/ and we can attempt to dig in more and try to replicate the same scenario you are experiencing if needed.

**Comment by jvliwanag** [comment]
Hi @orware ,

Thanks for looking into this.

While `PREPARE` over mysql shell works as it is sent over the MySQL QUERY command (code: 3), `PREPARE` as command (code: 22) itself sent over mysql drivers (as tried on rust and nodejs) exhibits this behavior. See sample nodejs repro gist [here](https://gist.github.com/jvliwanag/77fcaa57de04876f20d58d4f47945944).

Looking into wireshark traces of this.

Query/shell/dbeaver based MySQL PREPARE sends this (works):

<img width="932" alt="Screenshot 2024-01-18 at 10 04 33 AM" src="https://github.com/planetscale/discussion/assets/211068/c2ef211f-ad81-480a-9ac6-8e94b3ddd4b6">

Nodejs / rust driver based PREPARE command sends this (doesn't work):

<img width="814" alt="Screenshot 2024-01-18 at 10 05 00 AM" src="https://github.com/planetscale/discussion/assets/211068/49882805-8f44-45b3-86f5-f2a8ee3617d0">

**Reply by orware** [reply, marked as answer]
Hello @jvliwanag,

All of the extra information you provided is appreciated!

We had another user write in to support overnight about encountering the same issue and it allowed us to connect a few dots to better provide an explanation / possible solution for you here.

It does look like this is a `pscale connect` specific issue because using `pscale connect` still directs connections through our legacy edge routing infrastructure and work is still ongoing to migrate the command so that it uses our current edge routing infrastructure. This will likely arrive in the course of this year but until then this issue will likely remain.

The workaround however is quite simple, as using the standard hostname / credentials generated from within our UI today go through that newer edge routing infrastructure so if those are used instead you should not experience the issue. 

I believe you are already using the standard hostname / credentials successfully based on this comment: "Note that connecting directly to planetscale bypassing the pscale connect proxy works just fine." shared above in your initial message.

For now, if you are going to be relying on prepared statements within your application I would recommend sticking with using our provided hostname / credentials rather than using `pscale connect` due to this current limitation within that older routing infrastructure. Once `pscale connect` is eventually updated to utilize the newer routing infrastructure the issue will likely go away for it as well.

**Comment by jvliwanag** [comment]
Appreciate the time looking into this. Still very new with planetscale, aside from this very minor hiccup — planetscale already massively elevated what i now expect from a db provider. 😄 so awesome!

**Comment by mattrobenolt** [comment]
As an update, `pscale connect` has been drastically rewritten 5 days ago.

https://github.com/planetscale/cli/commit/56cfb502a63dc5b03db9f8bb81f05d82157e13e8

So it might fix your issue now? I'd be curious if it doesn't.
